Feat/fem attribute setters

aeverallpx opened the pull request:
Changes:
- Notebook for example of 3D topology optimisation with material orientation - we were working on this anyway but tried to build the example to match your 3D beam density optimisation as closely as possible.
Result:
![image](https://github.com/user-attachments/assets/54820baf-ec77-4e35-bb2f-d7de9ed4b2c8)

- property and setter in model forces, constraints and displacements. We had a bug where we set constraints as an Nx1 tensor rather than Nx3 but this failed silently and just returned bad optimisations. Hoping this change will catch any issues with this in future.

- TransverseIsotropicElasticity3D - light wrapper around OrthotropicElasticity3D to set parameter relationships.

aeverallpx committed:
Debug device and missing arguments

aeverallpx committed:
Add transversely isotropic elasticity

aeverallpx committed:
Merge pull request #1 from aeverallpx/feat/internal-solid-opt

Internal solid opt

aeverallpx committed:
Improve sparse class

aeverallpx committed:
coalesce K before returning

aeverallpx committed:
Merge pull request #2 from aeverallpx/feat/coalesce-bug

Coalesce bug

aeverallpx committed:
Setters and getters for FEM

aeverallpx committed:
Merge remote-tracking branch 'upstream/main' into feat/fem-attribute-setters

aeverallpx committed:
Topology optimisation notebook with 3d orientation

aeverallpx committed:
Material requirements

meyer-nils commented:
Thank you for adding this nice example. 

- The property and setter decorators are an excellent idea. What do you think about checking for types as well? For example, constraints should be always boolean.
- The wrapper for transversely isotropic materials is also a nice idea. Would you mind renaming `G` to `G_L`? In addition, there seems to be a factor 2 missing in the computation of `G_23`. At least, ABAQUS docs say so (https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/usb/default.htm?startat=pt05ch17s02abm02.html).
- The plot in your example re-implements some code from the already existing plot function for solids. I just added a rotation argument to the `Solid.plot()` function (accepting a rotation matrix) in a different branch and decided to commit it to main right away. Now, you can also pass in an existing plotter to make modifications like yours without duplicating the plot function. Would you mind trying to utilize that plot function in the example?
- Why did you uncomment `rtol` when calling minres?

aeverallpx committed:
Uncomment rtol

aeverallpx committed:
TransverseIsotropic fix G_23

aeverallpx committed:
Merge remote-tracking branch 'upstream/main' into feat/fem-attribute-setters

aeverallpx committed:
Plot from Solid.plot

aeverallpx committed:
Check data types in base

aeverallpx commented:
Thanks for the detailed comments. A few mistakes in there from me.

- Great suggestion, I've added type checks too!
- Thank you for catching this. I've corrected the TransverseIsotropic material class and changed to G_L.
- I'm now using Solid.plot(). I added an extra argument in there so that you can subset down to the cells which exceed the threshold as orientations of elements outside the structure aren't so interesting.
![image](https://github.com/user-attachments/assets/501471b4-8fc2-4cf0-b6ad-c2329aac8aad)
- That was a mistake from me - hang over from using the wrong scipy version before. I've re-uncommented it.

meyer-nils commented:
Great, thanks for the swift incorporation of the requested changes!
